bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/HadoopHashTableSource.java

```diff
@@ -0,0 +1,464 @@
+
+    final ImmutableBytesWritable startRowInclusive;
+    final ImmutableBytesWritable stopRowExclusive;
+
+    long numKeys = 0;
+    // Flag indicating that this workitem is finished.
+    boolean isDone = false;
+    ImmutableBytesWritable currentRangeStartKey;
+    // Hash for the current range.
+    ImmutableBytesWritable currentHash;
+    RangeHash currentRangeHash;
+
+    public HashBasedReader(
+        HadoopHashTableSource source,
+        ImmutableBytesWritable startRowInclusive,
+        ImmutableBytesWritable stopRowExclusive,
+        TableHashReader reader) {
+      this.reader = reader;
+      this.source = source;
+      this.startRowInclusive = startRowInclusive;
+      this.stopRowExclusive = stopRowExclusive;
+    }
+
+    @Override
+    public boolean start() throws IOException {
+      // NO CHECKED EXCEPTIONS HERE.
+      LOG.debug(
+          "Starting a new reader at key range ["
+              + immutableBytesToString(startRowInclusive)
+              + " ,"
+              + immutableBytesToString(stopRowExclusive)
+              + ").");
+      numKeys = 0;
+
+      if (readNextKey()) {
+        // Dataflow calls start, followed by getCurrent. HashBased reader needs to read on TableHash
+        // twice to return a RangeHash since it specifies both range-start and range-end.
+        advance();
+        return true;
+      }
+
+      isDone = true;
+      return false;
+    }
+
+    @Override
+    public boolean advance() throws IOException {
+      if (isDone) {
+        LOG.debug("Ending workitem at key " + immutableBytesToString(currentRangeStartKey) + " .");
+        return false;
+      }
+
+      ImmutableBytesWritable startKey = this.currentRangeStartKey;
+      ImmutableBytesWritable hash = this.currentHash;
+
+      if (!readNextKey()) {
+        this.currentRangeHash = RangeHash.of(startKey, stopRowExclusive, hash);
+        // return true since we have lastBatchStartKey to emit. Set isDone=true to prevent reading
+        // from a potentially exhausted reader.
+        isDone = true;
+      } else {
+        this.currentRangeHash = RangeHash.of(startKey, reader.getCurrentKey(), hash);
+      }
+
+      return true;
+    }
+
+    // Returns true if a key can be read for this workitem.
+    private boolean readNextKey() throws IOException {
+      if (reader.next()) {
+        numKeys++;
+        this.currentRangeStartKey = reader.getCurrentKey();
+        if ( // StopRow is not set, everything is in bounds.
+        (stopRowExclusive.equals(HConstants.EMPTY_END_ROW)
+            || currentRangeStartKey.compareTo(stopRowExclusive) < 0)) { // currentKey < stopKey
+          // There is a key to read and the key is within the bounds of this workitem. Return true.
+          this.currentHash = reader.getCurrentHash();
+          return true;
+        } else {
+          // There is a key to read but its outside of the bounds of this workitem.
+          this.currentHash = null;
```

Comment:
I didn't get the subrange part? We need to interpret the ranges.

Reply:
stopRowExclusive should never land in the middle of chunk, it should always equal the start of the next chunk (unless its the very last chunk). If doesnt, then that means that there is a bug somewhere else which will lead to under reading data in bigtable. I'm proposing to to proactively sanity check this invariant

Reply:
But we are always reading just one line. So are we checking for a bug in our own code? How does unit tests not do it? If we were reading a <start,end> tuple from HBase then it would have made sense that we are validating the end[i] == start[i+1] but we are just reading starts and creating the [start,end] ourselves, which is tested via unit/integ tests.